File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/submit_workflow_task_approval.go
package workflowexecution

import (
	"context"
	"time"

	"github.com/rs/zerolog/log"
	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
	"github.com/stigmer/stigmer/backend/libs/go/grpc/request/pipeline"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	wftemporal "github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/domain/workflowexecution/temporal"
	wfactivities "github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/domain/workflowexecution/temporal/activities"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/domain/workflowexecution/temporal/workflows"
)

const humanInputSignalPrefix = "human_input_"

// SubmitWorkflowTaskApproval submits a human reviewer's decision for a
// workflow-level human_input task.
//
// Constructs a Temporal signal matching the TS runner's expected
// HumanInputResult shape, wraps it in a relaySignal envelope, and
// delivers it via SignalWithStart to the Go outer workflow. The outer
// workflow's relaySignal handler forwards the signal to the TS child
// workflow where the human_input task is blocking.
//
// Pipeline Steps:
// 1. ValidateTaskApprovalInput - Validate required fields
// 2. LoadExecutionForApproval - Load execution from database
// 3. ValidateApprovalSignalable - Ensure execution is in a signalable phase
// 4. ValidateHumanInputTask - Ensure task exists and is a human_input task
// 5. SendTaskApprovalSignal - Build and send the relaySignal-wrapped signal
//
// @since T13c (Workflow HITL Approval UI)
func (c *WorkflowExecutionController) SubmitWorkflowTaskApproval(
	ctx context.Context,
	input *workflowexecutionv1.SubmitWorkflowTaskApprovalInput,
) (*workflowexecutionv1.WorkflowExecution, error) {
	log.Info().
		Str("execution_id", input.GetExecutionId()).
		Str("task_name", input.GetTaskName()).
		Str("outcome", input.GetOutcome()).
		Msg("SubmitWorkflowTaskApproval request")

	reqCtx := pipeline.NewRequestContext(ctx, input)

	p := c.buildSubmitWorkflowTaskApprovalPipeline()
	if err := p.Execute(reqCtx); err != nil {
		log.Warn().
			Str("execution_id", input.GetExecutionId()).
			Str("task_name", input.GetTaskName()).
			Err(err).
			Msg("SubmitWorkflowTaskApproval failed")
		return nil, err
	}

	execution := reqCtx.Get(LoadedExecutionKey)
	if execution == nil {
		return nil, grpclib.InternalError(nil, "execution not found in context after pipeline")
	}

	log.Info().
		Str("execution_id", input.GetExecutionId()).
		Str("task_name", input.GetTaskName()).
		Str("outcome", input.GetOutcome()).
		Msg("SubmitWorkflowTaskApproval completed")

	return execution.(*workflowexecutionv1.WorkflowExecution), nil
}

func (c *WorkflowExecutionController) buildSubmitWorkflowTaskApprovalPipeline() *pipeline.Pipeline[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput] {
	return pipeline.NewPipeline[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]("workflowexecution-submit-task-approval").
		AddStep(&ValidateTaskApprovalInputStep{}).
		AddStep(&LoadExecutionForApprovalStep{store: c.store}).
		AddStep(&ValidateApprovalSignalableStep{}).
		AddStep(&ValidateHumanInputTaskStep{}).
		AddStep(&SendTaskApprovalSignalStep{
			workflowCreator: c.workflowCreator,
			temporalConfig:  c.temporalConfig,
		}).
		Build()
}

// =============================================================================
// Step 1: Validate Input
// =============================================================================

type ValidateTaskApprovalInputStep struct{}

func (s *ValidateTaskApprovalInputStep) Name() string {
	return "ValidateTaskApprovalInput"
}

func (s *ValidateTaskApprovalInputStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]) error {
	input := ctx.Input()

	if input.GetExecutionId() == "" {
		return grpclib.InvalidArgumentError("execution_id is required")
	}
	if input.GetTaskName() == "" {
		return grpclib.InvalidArgumentError("task_name is required")
	}
	if input.GetOutcome() == "" {
		return grpclib.InvalidArgumentError("outcome is required")
	}

	return nil
}

// =============================================================================
// Step 2: Load Execution
// =============================================================================

type LoadExecutionForApprovalStep struct {
	store store.Store
}

func (s *LoadExecutionForApprovalStep) Name() string {
	return "LoadExecutionForApproval"
}

func (s *LoadExecutionForApprovalStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]) error {
	input := ctx.Input()
	executionID := input.GetExecutionId()

	existing := &workflowexecutionv1.WorkflowExecution{}
	if err := s.store.GetResource(ctx.Context(), apiresourcekind.ApiResourceKind_workflow_execution, executionID, existing); err != nil {
		return grpclib.NotFoundError("WorkflowExecution", executionID)
	}

	ctx.Set(LoadedExecutionKey, existing)
	return nil
}

// =============================================================================
// Step 3: Validate Signalable Phase
// =============================================================================

type ValidateApprovalSignalableStep struct{}

func (s *ValidateApprovalSignalableStep) Name() string {
	return "ValidateApprovalSignalable"
}

func (s *ValidateApprovalSignalableStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]) error {
	execution := ctx.Get(LoadedExecutionKey).(*workflowexecutionv1.WorkflowExecution)
	phase := execution.GetStatus().GetPhase()

	switch phase {
	case workflowexecutionv1.ExecutionPhase_EXECUTION_PENDING,
		workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS:
		return nil
	default:
		return grpclib.FailedPreconditionError(
			"cannot submit task approval: execution is in %s phase", phase.String(),
		)
	}
}

// =============================================================================
// Step 4: Validate Human Input Task
// =============================================================================

type ValidateHumanInputTaskStep struct{}

func (s *ValidateHumanInputTaskStep) Name() string {
	return "ValidateHumanInputTask"
}

func (s *ValidateHumanInputTaskStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]) error {
	input := ctx.Input()
	execution := ctx.Get(LoadedExecutionKey).(*workflowexecutionv1.WorkflowExecution)
	taskName := input.GetTaskName()

	tasks := execution.GetStatus().GetTasks()
	for _, task := range tasks {
		if task.GetTaskName() == taskName {
			if task.GetTaskType() != workflowexecutionv1.WorkflowTaskType_WORKFLOW_TASK_APPROVAL {
				return grpclib.InvalidArgumentError(
					"task '%s' is not a human_input task (type: %s)", taskName, task.GetTaskType().String(),
				)
			}
			return nil
		}
	}

	return grpclib.InvalidArgumentError(
		"task '%s' not found in execution status", taskName,
	)
}

// =============================================================================
// Step 5: Send Signal via relaySignal Wrapping
// =============================================================================

type SendTaskApprovalSignalStep struct {
	workflowCreator *workflows.InvokeWorkflowExecutionWorkflowCreator
	temporalConfig  *wftemporal.Config
}

func (s *SendTaskApprovalSignalStep) Name() string {
	return "SendTaskApprovalSignal"
}

func (s *SendTaskApprovalSignalStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.SubmitWorkflowTaskApprovalInput]) error {
	input := ctx.Input()
	execution := ctx.Get(LoadedExecutionKey).(*workflowexecutionv1.WorkflowExecution)
	executionID := input.GetExecutionId()

	// Reviewer attribution: OSS is single-user with no multi-tenant auth
	// context, so there is no authenticated principal to attribute the
	// decision to. An explicit client-supplied reviewer is honored (the CLI
	// may pass one); otherwise the reviewer stays empty — "Empty when not
	// attributed", matching the agent-execution ledger's decided_by contract.
	// The Cloud edition attributes server-side from the authenticated caller
	// and additionally stamps a reviewer_actor display snapshot.
	reviewer := input.GetReviewer()

	humanInputSignalName := humanInputSignalPrefix + input.GetTaskName()

	signalPayload := map[string]interface{}{
		"outcome":      input.GetOutcome(),
		"reviewer":     reviewer,
		"responded_at": time.Now().UTC().Format(time.RFC3339),
	}

	if input.GetFormData() != nil {
		signalPayload["form_data"] = input.GetFormData().AsMap()
	}

	if input.GetComment() != "" {
		signalPayload["comment"] = input.GetComment()
	}

	relayPayload := workflows.RelaySignalPayload{
		SignalName: humanInputSignalName,
		Payload:    signalPayload,
	}

	workflowInput := &wfactivities.InvokeWorkflowExecutionWorkflowInput{
		ExecutionID:        executionID,
		WorkflowInstanceID: execution.GetSpec().GetWorkflowInstanceId(),
		WorkflowID:         execution.GetStatus().GetTemporalWorkflowId(),
		OrgID:              execution.GetMetadata().GetOrg(),
	}

	if s.workflowCreator == nil {
		return grpclib.UnavailableError(
			"workflow creator not configured for task '%s'", input.GetTaskName(),
		)
	}

	dispatch := wftemporal.ResolveWorkflowTaskQueue(
		executionID,
		execution.GetSpec().GetExecutionTarget(),
		s.temporalConfig,
	)

	err := s.workflowCreator.SignalWithStart(
		ctx.Context(),
		workflowInput,
		"relaySignal",
		relayPayload,
		dispatch.TaskQueue,
	)
	if err != nil {
		return grpclib.UnavailableError(
			"failed to send approval signal for task '%s': %v", input.GetTaskName(), err,
		)
	}

	log.Info().
		Str("execution_id", executionID).
		Str("task_name", input.GetTaskName()).
		Str("outcome", input.GetOutcome()).
		Str("reviewer", reviewer).
		Str("signal_name", humanInputSignalName).
		Msg("AUDIT: Workflow task approval submitted via relaySignal")

	return nil
}

File: backend/services/stigmer-server/src/domain/workflowexecution/submit-workflow-task-approval.ts
/**
 * SubmitWorkflowTaskApproval — ports submit_workflow_task_approval.go
 * (T13c): a human reviewer's decision for a workflow-level human_input
 * task. Constructs the runner's HumanInputResult signal payload, wraps it
 * in the relaySignal envelope on the `human_input_{task}` channel, and
 * delivers via SignalWithStart (the orchestrator forwards to the TS child
 * where the task is blocking).
 *
 * Quirk ported faithfully: the SignalWithStart input's workflowId field
 * carries status.temporal_workflow_id (NOT spec.workflow_id — every other
 * sender passes the spec reference).
 *
 * Reviewer attribution: OSS is single-user; an explicit client-supplied
 * reviewer is honored, otherwise it stays empty ("Empty when not
 * attributed"). Cloud attributes server-side from the authenticated
 * caller.
 */
import type { JsonValue } from "@bufbuild/protobuf";

import type { WorkflowExecution } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { WorkflowExecutionSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { WorkflowExecutionCommandController } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/command_pb";
import {
  ExecutionPhase,
  WorkflowTaskType,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";
import type { SubmitWorkflowTaskApprovalInput } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../boot/logger.js";
import {
  failedPreconditionError,
  internalError,
  invalidArgumentError,
  notFoundError,
  unavailableError,
} from "../../pipeline/errors.js";
import { newPipeline } from "../../pipeline/pipeline.js";
import { RequestContext } from "../../pipeline/request-context.js";
import type { Store } from "../../store/interface.js";

import {
  HUMAN_INPUT_SIGNAL_PREFIX,
  RELAY_SIGNAL_CHANNEL_NAME,
} from "./constants.js";
import type { WorkflowExecutionEngineStateProvider } from "./engine.js";

export interface SubmitWorkflowTaskApprovalDeps {
  readonly store: Store;
  readonly logger: Logger;
  readonly engineState: WorkflowExecutionEngineStateProvider;
}

type TaskApprovalDesc =
  typeof WorkflowExecutionCommandController.method.submitWorkflowTaskApproval.input;

const LOADED_EXECUTION_KEY = "loadedExecution";

export async function submitWorkflowTaskApproval(
  deps: SubmitWorkflowTaskApprovalDeps,
  input: SubmitWorkflowTaskApprovalInput,
): Promise<WorkflowExecution> {
  const reqCtx = new RequestContext(
    WorkflowExecutionCommandController.method.submitWorkflowTaskApproval.input,
    input,
    ApiResourceKind.workflow_execution,
  );
  await newPipeline<TaskApprovalDesc>(
    "workflowexecution-submit-task-approval",
    deps.logger,
  )
    .addStep({
      name: "ValidateTaskApprovalInput",
      execute(ctx) {
        if (ctx.input.executionId === "") {
          throw invalidArgumentError("execution_id is required");
        }
        if (ctx.input.taskName === "") {
          throw invalidArgumentError("task_name is required");
        }
        if (ctx.input.outcome === "") {
          throw invalidArgumentError("outcome is required");
        }
      },
    })
    .addStep({
      name: "LoadExecutionForApproval",
      async execute(ctx) {
        let execution: WorkflowExecution;
        try {
          execution = await deps.store.getResource(
            ApiResourceKind.workflow_execution,
            ctx.input.executionId,
            WorkflowExecutionSchema,
          );
        } catch {
          throw notFoundError("WorkflowExecution", ctx.input.executionId);
        }
        ctx.set(LOADED_EXECUTION_KEY, execution);
      },
    })
    .addStep({
      name: "ValidateApprovalSignalable",
      execute(ctx) {
        const execution = ctx.get(LOADED_EXECUTION_KEY) as WorkflowExecution;
        const phase =
          execution.status?.phase ?? ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED;
        if (
          phase !== ExecutionPhase.EXECUTION_PENDING &&
          phase !== ExecutionPhase.EXECUTION_IN_PROGRESS
        ) {
          throw failedPreconditionError(
            `cannot submit task approval: execution is in ${ExecutionPhase[phase]} phase`,
          );
        }
      },
    })
    .addStep({
      name: "ValidateHumanInputTask",
      execute(ctx) {
        const execution = ctx.get(LOADED_EXECUTION_KEY) as WorkflowExecution;
        const taskName = ctx.input.taskName;
        const task = (execution.status?.tasks ?? []).find(
          (candidate) => candidate.taskName === taskName,
        );
        if (task === undefined) {
          throw invalidArgumentError(
            `task '${taskName}' not found in execution status`,
          );
        }
        if (task.taskType !== WorkflowTaskType.WORKFLOW_TASK_APPROVAL) {
          throw invalidArgumentError(
            `task '${taskName}' is not a human_input task (type: ${WorkflowTaskType[task.taskType]})`,
          );
        }
      },
    })
    .addStep({
      name: "SendTaskApprovalSignal",
      async execute(ctx) {
        const execution = ctx.get(LOADED_EXECUTION_KEY) as WorkflowExecution;
        const executionId = ctx.input.executionId;
        const humanInputSignalName =
          HUMAN_INPUT_SIGNAL_PREFIX + ctx.input.taskName;

        const signalPayload: Record<string, JsonValue> = {
          outcome: ctx.input.outcome,
          reviewer: ctx.input.reviewer,
          // Go time.Now().UTC().Format(time.RFC3339): seconds precision.
          responded_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
        };
        if (ctx.input.formData !== undefined) {
          signalPayload["form_data"] = ctx.input.formData as JsonValue;
        }
        if (ctx.input.comment !== "") {
          signalPayload["comment"] = ctx.input.comment;
        }

        const engineState = deps.engineState();
        if (!engineState.connected) {
          throw unavailableError(
            `workflow creator not configured for task '${ctx.input.taskName}'`,
          );
        }
        try {
          await engineState.engine.signalWithStart(
            {
              executionId,
              workflowInstanceId: execution.spec?.workflowInstanceId ?? "",
              // Go's quirk: the temporal workflow id, not spec.workflow_id.
              workflowId: execution.status?.temporalWorkflowId ?? "",
              orgId: execution.metadata?.org ?? "",
              recoveryMode: false,
              executionTarget: execution.spec?.executionTarget ?? 0,
            },
            RELAY_SIGNAL_CHANNEL_NAME,
            { signalName: humanInputSignalName, payload: signalPayload },
          );
        } catch (error) {
          throw unavailableError(
            `failed to send approval signal for task '${ctx.input.taskName}': ${error instanceof Error ? error.message : String(error)}`,
          );
        }
        deps.logger.info(
          "AUDIT: Workflow task approval submitted via relaySignal",
          {
            executionId,
            taskName: ctx.input.taskName,
            outcome: ctx.input.outcome,
            signalName: humanInputSignalName,
          },
        );
      },
    })
    .build()
    .execute(reqCtx);

  const execution = reqCtx.get(LOADED_EXECUTION_KEY);
  if (execution === undefined) {
    throw internalError(
      new Error("execution not found in context after pipeline"),
      "execution not found in context after pipeline",
    );
  }
  return execution as WorkflowExecution;
}
